apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -226,45 +219,44 @@ public sealed class KeyRange
             EndInclusive = endInclusive;
         }
 
-        /// <inheritdoc/>
-        public override bool Equals(object other)
-            => other is KeyRange range && Protobuf.Equals(range.Protobuf);
-
-        /// <inheritdoc/>
-        public override int GetHashCode() => Protobuf.GetHashCode();
-
-        /// <inheritdoc/>
-        public override string ToString()
+        /// <summary>
+        /// Returns the string representation of this KeyRange using the type conversion options specified in 
+        /// <see cref="SpannerConnectionStringBuilder"/>. The string representation is only for diagnostic purposes
+        /// and the implementation may change over time.
+        /// </summary>
+        /// <param name="builder">The <see cref="SpannerConnectionStringBuilder"/> instance.</param>
+        /// <returns>A string that represents the current KeyRange.</returns>
+        public string ToString(SpannerConnectionStringBuilder builder)
         {
-            var builder = new StringBuilder();
-            builder.Append(BeginInclusive ? '[' : '(');
-            builder.Append(BeginAt.KeyParts);
-            builder.Append(", ");
-            builder.Append(End.KeyParts);
-            builder.Append(EndInclusive ? ']' : ')');
-
-            return builder.ToString();
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append(BeginInclusive ? '[' : '(');
+            stringBuilder.Append(BeginAt.ToString(builder));
+            stringBuilder.Append(", ");
+            stringBuilder.Append(End.ToString(builder));
+            stringBuilder.Append(EndInclusive ? ']' : ')');
+
+            return stringBuilder.ToString();
         }
 
-        private V1.KeyRange ToProtobuf()
+        internal V1.KeyRange ToProtobuf(SpannerConversionOptions options)
         {
             var range = new V1.KeyRange();
             if (BeginInclusive)
             {
-                range.StartClosed = BeginAt.KeyParts;
+                range.StartClosed = BeginAt.ToProtobuf(options);
```

Comment:
Another refactoring: I'd use an expression-bodied member here assigning all properties, just assigning null values where necessary. Basically remind me to have a look at this file later :)